cirq/protocols/apply_channel.py

```diff
@@ -12,3 +12,320 @@
+# be of type np.ndarray to ensure the method has the correct type signature in
+# that case. It is checked for using `is`, so it won't have a false positive if
+# the user provides a different np.array([]) value.
+
+RaiseTypeErrorIfNotProvided = np.array([])  # type: np.ndarray
+
+TDefault = TypeVar('TDefault')
+
+
+class Axes:
+    """An enum for the two types of axes, row and column indices.
+
+    For a tensor with shape (d_1_,...d_k), some subset of the first
+    of these indices correspond to row indices, and the rest of these
+    indices correspond to column indices.  These are row and column, in the
+    sense that they describe a matrix where the row indices could be
+    reshaped to form a row, and similar for the column indices.
+    """
+    ROW = 1
+    COLUMN = 2
+
+
+class ApplyChannelArgs:
+    r"""Arguments for efficiently performing a channel.
+
+    A channel performs the mapping
+        $$
+        X \rightarrow \sum_i A_i X A_i^\dagger
+        $$
+    for operators $A_i$ that satisfy the normalization condition
+        $$
+        \sum_i A_i^\dagger A_i = I.
+        $$
+
+    The receiving object is expected to mutate `target_tensor` so that it
+    contains the density matrix after multiplication, and then return
+    `target_tensor`. Alternatively, if workspace is required,
+    the receiving object can overwrite `out_buffer` with the results
+    and return `out_buffer`. Or, if the receiving object is attempting to
+    be simple instead of fast, it can create an entirely new array and
+    return that.
+
+    Attributes:
+        target_tensor: The input tensor that needs to be left and right
+            multiplied and summed, representing the effect of the channel.
+            The tensor will have the shape (2, 2, 2, ..., 2). It usually
+            corresponds to a multi-qubit density matrix, with the first
+            n indices corresponding to the rows of the density matrix and
+            the last n indices corresponding to the columns of the density
+            matrix.
+        out_buffer: Pre-allocated workspace with the same shape and
+            dtype as the target tensor. If buffers are used, the result should
+            end up in this buffer. It is the responsibility of calling code
+            to notice if the result is this buffer.
+        auxiliary_buffer0: Pre-allocated workspace with the same shape and dtype
+            as the target tensor.
+        auxiliary_buffer1: Pre-allocated workspace with the same shape
+            and dtype as the target tensor.
+        left_axes: Which axes to multiply the left action of the channel upon.
+        right_axes: Which axes to multiply the right action of the channel upon.
+    """
+
+    def __init__(self,
+            target_tensor: np.ndarray,
+            out_buffer: np.ndarray,
+            auxiliary_buffer0: np.ndarray,
+            auxiliary_buffer1: np.ndarray,
```

Comment:
\sum_k A_k \rho A_k^\dagger

out is for accumulation of the sum 
aux buf0 is because you need to cache the initial state
aux buf1 is because you need to do efficient multiplications between this buf and target

Reply:
I see.

It would be preferable if we didn't distinguish between the buffers, and the user was free to return any of them. There's no real distinction between `out_buffer` and `auxiliary_buffer0` other than that your program silently does bad things if you put the result in `auxiliary_buffer0`.

Reply:
sorry I don't follow what you are saying.  out_buffer is distinguished by being the only one that can recieve the result.  If you do it your way the caller of this will have to check each of these separately to see which "is" the result.  That feels bad to me.

Reply:
Discussed off github.  Will keep output in out buffer but make assertion to catch problems.